### DISCUSSION TITLE
Restarts at t = 0h

### DISCUSSION THREAD

**Discussion by tomgowan** [ufs-community/ufs-weather-model#1160] [Q&A] [answered]
Is there a way to generate restart files at t =0h (i.e., immediately after model initialization)? I've tried setting `restart_interval = 0,-1` and restart_interval = 24,-1` with no success. Thanks.

**Comment by pjpegion** [comment]
Someone else might know more specifics, but it is my understanding that the code to write the restarts is at the end of the time-loop, so the best you can do is run the model one time-step and write out the restarts. I just checked the model code, and nhours_fcst can be a fraction of an hour (e.g 0.25  if you are running with a 900 second time-step)

**Reply by tomgowan** [reply]
Thanks @pjpegion . We have been able to do this in MPAS and WRF (WRF required some source code changes). I'm currently testing if something can be added to `FV3/module_fcst_grid_comp.F90` that will force UFS to generate these restarts at t-0h. If you are correct that the first check to generate restarts is at the end of the first time-loop, this may not be doable.

**Comment by jiandewang** [comment]
for MOM, this is what you need in MOM_input
SAVE_INITIAL_CONDS = True      !   [Boolean] default = False
                                ! If true, write the initial conditions to a file given by IC_OUTPUT_FILE.
I think CICE will save IC inside history directory

**Reply by tomgowan** [reply]
Thanks @jiandewang. I should have been more specific in that I'm trying to do this in UFS/FV3. It doesn't look like this option exists there. Also, would this option generate restart files or just standard output files?

**Reply by jiandewang** [reply]
> Thanks @jiandewang. I should have been more specific in that I'm trying to do this in UFS/FV3. It doesn't look like this option exists there. Also, would this option generate restart files or just standard output files?

for MOM it will simply generate IC, not restart file

**Comment by tomgowan** [comment, marked as answer]
I was able to generate restarts by editing the source code file **FV3/module_fcst_grid_comp.F90**. I first simply forced FV3 to generate restarts files every time step within the **fcst_run_phase_2** subroutine by removing the if statements that checked if restarts should be generated based on the **restart_interval** parameter in **model_configure**.  As @pjpegion correctly assumed, this resulted in the first set of restart files being generated after the first time step. 

I then decided to move this code into the preceding subroutine named **fcst_run_phase_1** (before any integration occurs). Somewhat surprisingly, this worked and restart files were generated at t=0. There do not appear to be any issues with the files and I was able to restart a run using them.

Here's my final implementation within **fcst_run_phase_1**:
*note that a few variables shown here need to be added to the subroutine

 ```
737 !-------------------------------------------------------------------------
 738 !********************** START MY CHANGES *****************************
 739 !-------------------------------------------------------------------------
 740 
 741 ! This code block was added to generate restart files at t=0 (just after
 742 ! intialization). The default location to generate restart files is in the
 743 ! subroutine below called "fcst_run_phase_2". However, this subroutine is
 744 ! first called after the first model time step, and thus, will not allow
 745 ! for restart files to be generated at t=0. Copying the calls to generate 
 746 ! restarts here, in the subroutine called "fcst_run_phase_1", allows us 
 747 ! to generate restart files at t=0. 
 748 
 749 ! Flag to generate restarts at t=0.
 750 restart_t0 = .true.
 751 
 752 ! Get current time in seconds
 753 call get_time(atm_int_state%Time_atmos - atm_int_state%Time_atstart, seconds)
 754 
 755 if (restart_t0 == .true.) then
 756   if (seconds == 0) then
 757 
 758     ! Generate restart files. Same as default block of code in
 759     ! subroutine below called "fcst_run_phase_2".   
 760     call get_time(atm_int_state%Time_atmos - atm_int_state%Time_atstart, seconds)
 761     restart_inctime = set_time(seconds, 0)
 762     atm_int_state%Time_restart = atm_int_state%Time_atstart + restart_inctime
 763     timestamp = date_to_string (atm_int_state%Time_restart)
 764     call atmos_model_restart(atm_int_state%Atm, timestamp)
 765     call wrt_atmres_timestamp(atm_int_state,timestamp)
 766     print *, 'Restart files at t = 0 seconds generated.'
 767 
 768   endif
 769 endif
 770 
 771 !-------------------------------------------------------------------------
 772 !************************ END MY CHANGES *****************************
 773 !-------------------------------------------------------------------------
```
